keel-web/src/test/kotlin/com/netflix/spinnaker/keel/rest/AuthorizationTestSupport.kt

```diff
@@ -0,0 +1,50 @@
+package com.netflix.spinnaker.keel.rest
+
+import com.fasterxml.jackson.databind.ObjectMapper
+import dev.minutest.ContextBuilder
+import io.mockk.every
+import org.springframework.http.MediaType
+import org.springframework.test.web.servlet.MockMvc
+import org.springframework.test.web.servlet.request.MockMvcRequestBuilders.delete
+import org.springframework.test.web.servlet.request.MockMvcRequestBuilders.get
+import org.springframework.test.web.servlet.request.MockMvcRequestBuilders.post
+import org.springframework.test.web.servlet.request.MockMvcRequestBuilders.put
+import org.springframework.test.web.servlet.result.MockMvcResultMatchers.status
+
+fun ContextBuilder<*>.testApiPermissions(
+  mvc: MockMvc,
+  jsonMapper: ObjectMapper,
+  authorizationSupport: AuthorizationSupport,
+  apis: Map<ApiRequest, AuthorizationSupport.Permission>
+) {
+  apis.forEach { (api, permission) ->
+    val (method, uri) = api.request.split(Regex("""\s+"""))
+    context("API permission check for $api") {
+      before {
+        every {
+          authorizationSupport.userCan(permission.action.name, permission.entity.name, any())
+        } returns false
+      }
```

Comment:
@robfletcher This is what I ended up with that worked, but was not what I originally intended. What I would like to do is to have a single context like "API permission checks" for the entire controller, with multiple tests underneath, one per API check. However, what I found is that trying to override the mock behavior in the `before` block inside each test does not work.

If you can think of something I'm missing in setting this up, please let me know.

Reply:
I don't understand. There's only one test in here, and I don't see anything overriding what this mock does.

Reply:
The original code was not generalized in this function, it was just in `ApplicationControllerTests`, and looked something like this:
```kotlin
  fun tests() = rootContext<Fixture> {
    fixture { Fixture() }

    before {
      every { authorizationSupport.userCan(Action.READ.name, Entity.APPLICATION.name, any()) } returns true      
    }

    after {
      clearAllMocks()
    }

    context("other controller tests") {
      // ...
    }

    context("when user is not authorized") {
      test("Some GET API is forbidden") {
        before {
          every { authorizationSupport.userCan(Action.READ.name, Entity.APPLICATION.name, any()) } returns true      
        }
        // mvc.perform...
      }

      test("Some POST API is forbidden") {
        before {
          every { authorizationSupport.userCan(Action.WRITE.name, Entity.APPLICATION.name, any()) } returns true      
        }
        // mvc.perform...
      }
    }
  }
```

This did not work.

Reply:
Why not wrap the cases where the user _is_ authorized in their own context?

Also you *cannot* put a `before` inside a `test` like that. I don't understand what the point is anyway. The point of grouping things into contexts is so they can share common setup. Why does each test in that context need its own special `before` block? Just generalize the mock initialization. Either:

```kotlin
every { authorizationSupport.userCan(Action.READ.name, Entity.APPLICATION.name, any()) } returns true 
every { authorizationSupport.userCan(Action.WRITE.name, Entity.APPLICATION.name, any()) } returns true
```

or…

```kotlin
every { authorizationSupport.userCan(any(), Entity.APPLICATION.name, any()) } returns true 
```

Reply:
> Also you **cannot** put a `before` inside a `test` like that.

Ohhh! There you go! So that's what I was missing. This feels like a bug that the DSL allows you to even do that.

Thanks for the example. In my mind, besides sharing common setup, another reason to use contexts was simply to logically organize your tests in a hierarchical way. But point taken! Will revisit.

Reply:
I'm not sure how it could prevent you doing that. It's kind of a limitation of the fact that even when a closure overrides `this` the `this` from any outer context is still visible.

Totally, organizing according to a logical hierarchy is the point, but I don't see that as being different from common setup. Usually the one requires the other, no?

Reply:
I've revisited the implementation here, please take a look: https://github.com/spinnaker/keel/pull/955/commits/09bd23dc3325c896c670947a3cd22b88b186dc69
Looks like this for `ApplicationControllerTests`:
![image](https://user-images.githubusercontent.com/1323478/78180786-219a0300-7418-11ea-9d2f-89de4072821c.png)